openssl/src/ocsp.rs

```diff
@@ -345,3 +417,56 @@ foreign_type_and_impl_send_sync! {
     pub struct OcspOneReq;
     pub struct OcspOneReqRef;
 }
+
+pub fn check_nonce(req: &OcspRequestRef, bs: &OcspBasicResponseRef) -> Result<OcspNonceCheckSuccessResult, OcspNonceCheckErrorResult> {
```

Comment:
I'd have this be a bit simpler:

```rust
pub fn check_nonce(req: &OcspRequestRef, bs: &OcspBasicResponseRef) -> NonceCheck { ... }

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct NonceCheck(c_int);

impl NonceCheck {
    pub const REQUEST_ONLY: NonceCheck = NonceCheck(-1);
    pub const UNEQUAL: NonceCheck = NonceCheck(0);
    pub const EQUAL: NonceCheck = NonceCheck(1);
    pub const ABSENT: NonceCheck = NonceCheck(2);
    pub const RESPONSE_ONLY: NonceCheck = NonceCheck(3);
}
```

Reply:
I agree that my version is complicated, but it has the benefit of making the user handle the error case; I believe that failing to `if (thing_that_returns_nonzero_on_error()) { goto uh_oh; }` is a relatively common source of security bugs. In my version, the compiler will warn about an unused result or if the user chose to use `unwrap()`/`?` it'll panic/propagate if proper error handling isn't done.

I'm happy to use your version if you disagree, but I think my version gives a more ergonomic usage for the common case.

(Warning: pseudocode)

```rust
fn perform_ocsp_request(...) -> ... {
    let my_req = ocsp::OcspRequest::new().unwrap();
    let res = send_request(&my_req);

    ocsp::check_nonce(&my_req, &res)?
}
```

Versus

```rust
fn perform_ocsp_request(...) -> ... {
    let my_req = ocsp::OcspRequest::new().unwrap();
    let res = send_request(&my_req);

    match ocsp::check_nonce(&my_req, &res) {
        ocsp::NonceCheck::REQUEST_ONLY => Err(()),
        ocsp::NonceCheck::UNEQUAL => Err(()),
        ocsp::NonceCheck::EQUAL => Ok(()),
        ...
    }
}
```

It only moves the work of handling the error up a level, but at least it's impossible to avoid handling it somewhere.

Looking forward to hearing your thoughts on this.

Reply:
How unambiguous is the error/ok decision in this context?

We can always tag the return type here `#[must_use]` just like `Result`.

Reply:
That's a good point. The OpenSSL docs make a reasonable case that at least a 0 return value should be an unambiguous error signal:

>The return values of OCSP_check_nonce() can be checked to cover each case. A positive return value effectively indicates success: nonces are both present and match, both absent or present in the response only. A non-zero return additionally covers the case where the nonce is present in the request only: this will happen if the responder doesn't support nonces. A zero return value indicates present and mismatched nonces: this should be treated as an error condition.

https://www.openssl.org/docs/man1.1.0/man3/OCSP_check_nonce.html

I like the idea of using `#[must_use]` for this at least, but I could see an argument that -1 at least shouldn't exactly be an error. Maybe a combination of our approaches like

```rust
pub fn check_nonce(req: &OcspRequestRef, bs: &OcspBasicResponseRef) -> Result<NonceCheck, SomeErrorNotSureWhat> { ... }

impl NonceCheck {
    pub const REQUEST_ONLY: NonceCheck = NonceCheck(-1);
    pub const EQUAL: NonceCheck = NonceCheck(1);
    pub const ABSENT: NonceCheck = NonceCheck(2);
    pub const RESPONSE_ONLY: NonceCheck = NonceCheck(3);
}
```